@here/harp-webtile-datasource/lib/HereWebTileDataSource.ts

```diff
@@ -0,0 +1,424 @@
+    ): Promise<CopyrightInfo[]> {
+        if (this.m_options.gatherCopyrightInfo === false) {
+            return [this.HERE_COPYRIGHT_INFO];
+        }
+
+        this.m_copyrightProvider.setRequestHeaders(requestHeaders);
+        return this.m_copyrightProvider.getCopyrights(tile.geoBox, tile.tileKey.level);
+    }
+
+    private parseBaseUrl(url: string): MapTileParams {
+        const parsed = new URL(url.startsWith("https:") ? url : `https://${url}`);
+        const fullPath = parsed.pathname;
+        const maptilePathRegexp = new RegExp("^(/maptile/2.1/)([^/]+)/([^/]+)/([^/]+)");
+        const match = fullPath.match(maptilePathRegexp);
+        if (!match) {
+            throw new Error(`HereWebTileDataSource: invalid baseUrl: ${url}`);
+        }
+        return {
+            baseUrl: parsed.host,
+            path: match[1],
+            tileType: match[2],
+            mapVersion: match[3],
+            scheme: match[4]
+        };
+    }
+
+    private getAuthParams(): string[] {
+        const { apikey } = this.m_options as ApiKeyAuthentication;
+        const { appId, appCode } = this.m_options as AppIdAuthentication;
+        const { authenticationCode } = this.m_options as TokenAuthentication;
+
+        const useAuthenticationCode = authenticationCode !== undefined;
+        const useApiKey = apikey !== undefined;
+        const useAppId = appId !== undefined && appCode !== undefined;
+
+        if (useAuthenticationCode) {
+            return [];
+        } else if (useApiKey) {
+            return [`apikey=${apikey}`];
+        } else if (useAppId) {
+            return [`app_id=${appId}`, `app_code=${appCode}`];
+        }
+
+        throw new Error("Neither apiKey, appId/appCode nor authenticationCode are defined.");
+    }
+
+    private getCopyrightRequestParams(): string {
+        const requestParams = ["output=json", ...this.getAuthParams()];
+
+        return `?${requestParams.join("&")}`;
+    }
+
+    private getImageRequestParams(): string {
+        const requestParams = this.getAuthParams();
+
+        if (this.m_options.additionalRequestParameters !== undefined) {
+            requestParams.push(this.m_options.additionalRequestParameters);
+        }
+        if (this.m_ppi !== WebTileDataSource.ppiValue.ppi72) {
+            // because ppi=72 is default, we do not include it in the request
+            requestParams.push(`ppi=${this.m_ppi}`);
+        }
+        if (this.m_languages !== undefined && this.m_languages[0] !== undefined) {
+            requestParams.push(`lg=${this.m_languages[0]}`);
+        }
+
+        if (this.m_languages !== undefined && this.m_languages[1] !== undefined) {
+            requestParams.push(`lg2=${this.m_languages[1]}`);
+        }
+
+        if (requestParams.length > 0) {
+            return `?${requestParams.join("&")}`;
+        }
+
+        return "";
+    }
+
+    public mapIsoLanguageToWebTile(languages: string[]): void {
```

Comment:
I think the private methods need to come at the bottom? I guess the bot will complain?

Reply:
done